File: DeductionServer/DeductionServer/MatchmakerHandler.cpp
#include "MatchmakerHandler.h"

#include "NetworkHandler.h"

MatchmakerHandler::MatchmakerHandler()
{
	link.handler = this;
	link.Open(asio::ip::udp::endpoint(asio::ip::udp::v4(), 0));
	link.Receive();
}

void MatchmakerHandler::UpdateLobby()
{
	lobby = "";
	link.Connect(asio::ip::udp::endpoint(asio::ip::address::from_string("172.105.79.194"), 58712));
}

void MatchmakerHandler::AcceptHandler(const asio::ip::udp::endpoint & endpoint)
{
}

void MatchmakerHandler::ConnectHandler(const asio::ip::udp::endpoint & endpoint)
{
	LobbyIdentity message;
	message.port = handler->link.socket.local_endpoint().port();
	link.Send(endpoint, message);
}

void MatchmakerHandler::LobbyIdentityHandler(const asio::ip::udp::endpoint & endpoint, const LobbyIdentity & message)
{
	lobby = message.lobby;
	std::cout << lobby << std::endl;
}

void MatchmakerHandler::LobbyRequestHandler(const asio::ip::udp::endpoint & endpoint, const LobbyRequest & message)
{
}

File: Assets/Scripts/Matchmaker/MatchmakerHandler.cs
using System;
using System.Net;

public class MatchmakerHandler
{
    public GameController game;

    public MatchmakerLink link;

    public string address = "172.105.79.194";
    public int port = 58712;
    public string lobby;

    public MatchmakerHandler()
    {
        link = new MatchmakerLink();
        link.handler = this;
        link.Open(new IPEndPoint(IPAddress.Any, 0));
        link.Receive();
    }

    public void ConnectToLobby(string lobby)
    {
        this.lobby = lobby;
        link.Connect(new IPEndPoint(IPAddress.Parse(address), port));
        game.connectionState = GameController.ConnectionState.ConnectingToMatchmaker;
        game.timeout = game.time + 2000000000;
    }

    internal void LobbyIdentityHandler(IPEndPoint endpoint, LobbyIdentity message)
    {
        if (message.address.Length == 0)
        {
            game.connectionState = GameController.ConnectionState.None;
            game.timeout = 0;
            if (message.lobby.Length == 0)
                game.CreateInfoPopup("No vacant lobbies.");
            else
                game.CreateInfoPopup("Lobby not found.");
        }
        else
        {
            if (lobby.Length == 0)
                lobby = message.lobby;
            if (lobby != message.lobby)
                game.CreateInfoPopup("Lobby mismatch.");
            game.handler.link.Connect(new IPEndPoint(IPAddress.Parse(message.address), message.port));
            game.connectionState = GameController.ConnectionState.ConnectingToLobby;
            game.timeout = game.time + 2000000000;
        }
    }

    internal void LobbyRequestHandler(IPEndPoint endpoint, LobbyRequest message)
    {
    }

    internal void ConnectionHandler(IPEndPoint endpoint)
    {
        link.Send(endpoint, new LobbyRequest { lobby = lobby });
        game.connectionState = GameController.ConnectionState.RequestingLobby;
        game.timeout = game.time + 2000000000;
    }
}
